### ISSUE TITLE
ScheduledReporter#stop() throws uncaught CancellationException

### ISSUE THREAD

**Issue by garsEstTemps** [dropwizard/metrics#1168]
### Given:

The following code which creates a new GraphiteReporter using a custom executor service which we not want to stop when the reporter should be stoped.

```java
final PickledGraphite graphite = new PickledGraphite(new InetSockedAddress("host", port));
this.graphiteReporter = 
GraphiteReporter.forRegistry(getTelemetryMetricsRegistry())
                               .scheduleOn(myCustomExecutorService)
                               .shutdownExecutorOnStop(false)
                               .build(graphite);
```

### When:
Stopping the reporter

```java
this.graphiteReporter.stop();
```

### Then:
We got an uncaught `CancellationException` thrown by `java.util.concurrent.FutureTask.java`:

```java
    private V report(int s) throws ExecutionException {
        Object x = outcome;
        if (s == NORMAL)
            return (V)x;
        if (s >= CANCELLED)
            throw new CancellationException();
        throw new ExecutionException((Throwable)x);
    }
```

The exception is thrown because the `ScheduledReporter` is calling `FutureTask#get()` after calling `FutureTask#cancel()` and expects an `ExecutionException` to be thrown but not a `CancellationException`:

```java
public void stop() {
        if (shutdownExecutorOnStop) {
            [...] // shutdownExecutorOnStop is always false in our case
        } else {
            // The external manager(like JEE container) responsible for lifecycle of executor
            synchronized (this) {
                if (this.scheduledFuture == null) {
                    // was never started
                    return;
                }
                if (this.scheduledFuture.isCancelled()) {
                    // already cancelled
                    return;
                }
                // just cancel the scheduledFuture and exit
                this.scheduledFuture.cancel(false);
                try {
                    // Wait a while for existing tasks to terminate
                    scheduledFuture.get(1, TimeUnit.SECONDS);
                } catch (ExecutionException e) {
                    // well, we should get this error when future is cancelled normally, just ignore it
                } catch (InterruptedException e) {
                    // The thread was interrupted while waiting future to complete
                    // Preserve interrupt status
                    Thread.currentThread().interrupt();
                    if (!this.scheduledFuture.isDone()) {
                        LOG.warn("The reporting schedulingFuture is not cancelled yet");
                    }
                } catch (TimeoutException e) {
                    // The last reporting cycle is still in progress, nothing wrong, just add log record
                    LOG.warn("The reporting schedulingFuture is not cancelled yet");
                }
            }
        }
```

Maybe a solution would be to modify the first catch clause to include the `CancellationException`:
```java
[...]
                try {
                    // Wait a while for existing tasks to terminate
                    scheduledFuture.get(1, TimeUnit.SECONDS);
                } catch (ExecutionException | CancellationException e) {
                    // well, we should get this error when future is cancelled normally, just ignore it
                }
[...]
```

**Comment by vladimir-bukhtoyarov** [comment]
@gboismal yes, you are right. CancellationException should be catched as well. 
It was my mistake. As explicitly documented in javadocs for [ScheduledExecutorService#scheduleAtFixedRate](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleAtFixedRate(java.lang.Runnable,%20long,%20long,%20java.util.concurrent.TimeUnit))
>  Returns:
a ScheduledFuture representing pending completion of the task, and whose get() method will throw an exception upon cancellation

As dirty workaround you can catch this exception inside your code, and remove this workaround when fix will be released.

**Comment by vladimir-bukhtoyarov** [comment]
I`ll prepare fix for 3.2 and 4.0 branches. 

The escalation reason is using mock instead of real executor service in [ScheduledReporterTest line 35](https://github.com/dropwizard/metrics/blob/4.0-development/metrics-core/src/test/java/com/codahale/metrics/ScheduledReporterTest.java#L35). I am going to replace mock by real executor for several test methods of ScheduledReporterTest, does anybody has objections?

**Comment by arteam** [comment]
No, sounds like a good plan.

**Comment by vladimir-bukhtoyarov** [comment]
Here bugfix https://github.com/dropwizard/metrics/pull/1170

**Comment by arteam** [comment]
Closed via #1170
